collada/triangleset.py

```diff
@@ -397,3 +413,31 @@ def __str__(self):
     def __repr__(self):
         return str(self)
 
+
+def _extendFromStrip(indexlist, index):
+    """Convert triangle strip indices to triangle indices
+
+    :param list indexlist:
+      flat triangles index list to append items to
+    :param numpy.ndarray index:
+      (#vertices, #inputs) shaped index array
+    """
+    for i in xrange(len(index) - 2):
+        if i % 2 == 0:
+            # clockwise
+            indexlist.extend(index[i:i + 3].flat)
+        else:
+            # counterclockwise
+            indexlist.extend(index[i + 2:i - 1:-1].flat)
+
+def _extendFromFan(indexlist, index):
+    """Convert triangle fan indices to triangle indices
+    """
+    for i in xrange(1, len(index) - 1):
```

Comment:
I'd like to avoid loops like this whenever possible, because it's very slow when loading large meshes.

For example, you could rewrite this with numpy operations. Let's say you have this list of trifan:

`>>> v = np.array([0,0, 1,3, 2,1,  3,2,  3,-2,  1,-3])
`
That's 4 triangles. The expanded indices would look like this:

`>>> e = np.array([0,0,1,3,2,1,  0,0,2,1,3,2,  0,0,3,2,3,-2,  0,0,3,-2,1,-3])
`
To build that array from the previous one, something like this works:

```
>>> v = v.reshape(6,2)
>>> v0 = np.repeat(v[0], len(additional_pts)).reshape(len(additional_pts), 2)
>>> prev_pts = v[2:-1]
>>> additional_pts = v[3:]
>>> np.concatenate((v[:3], np.vstack((v0, prev_pts, additional_pts))))
array([[ 0,  0],
       [ 1,  3],
       [ 2,  1],
       [ 0,  0],
       [ 0,  0],
       [ 0,  0],
       [ 2,  1],
       [ 3,  2],
       [ 3, -2],
       [ 3,  2],
       [ 3, -2],
       [ 1, -3]])
```

This ends up being much much more efficient than the loop you have here for large arrays. Could you try rewriting the _extendFromFan and _extendFromStrip functions?

Reply:
have you done any timing? Honestly, this sounds like premature optimization to me. The current for-loop is very readable and concise. Also, triangle strips and fans in real-world examples are typically sets of shorter strips and fans series. Consider this example: [duck_tristrips.dae](https://github.com/KhronosGroup/COLLADA-CTS/blob/master/StandardDataSets/collada_other/library_geometries/geometry/mesh/tristrips/duck_tristrips/duck_tristrips.dae)

I see a minor speedup with this implementation:

```python
# on line 223: index = numpy.concatenate(indexlist)

def _extendFromStrip(indexlist, index):
    cw_ = (index[0:-2:2], index[1:-1:2], index[2::2])
    ccw = (index[2:-1:2], index[1:-2:2], index[3::2])
    indexlist.append(numpy.swapaxes(cw_, 0, 1).reshape(-1))
    indexlist.append(numpy.swapaxes(ccw, 0, 1).reshape(-1))

def _extendFromFan(indexlist, index):
    c = numpy.concatenate((
        numpy.repeat(index[:1], len(index) - 2, 0),
        index[1:-1],
        index[2:]), 1)
    indexlist.append(c.reshape(-1))
```

Let me know if this is really what you prefer.

Reply:
branch with proposed no-loop code: [tristrips-no-loops](https://github.com/speleo3/pycollada/compare/tristrips...speleo3:tristrips-no-loops)

Reply:
I agree it might be premature, but I spent a lot of time optimizing triangls and polylist, so I want to make sure we have a similar speed for these two new primitives. You said there was a minor speedup: how did you test it? Was it with the duck_tristrips.dae file? I'm curious if you benchmarked it.

Reply:
ping? I've been asked to get a new release of pycollada out to pick up some fixes, but I was waiting for this to go in

Reply:
Sorry for the late reply, here are some timings.

Files from:
https://github.com/KhronosGroup/COLLADA-CTS/tree/master/StandardDataSets/1_5/collada_other/library_geometries

Branches:
https://github.com/speleo3/pycollada/tree/tristrips
https://github.com/speleo3/pycollada/tree/tristrips-no-loops

Script:
```
import collada, glob, os, timeit, subprocess, sys

p = subprocess.Popen(['git', 'rev-parse', '--abbrev-ref', 'HEAD'], stdout=subprocess.PIPE)
branch = p.communicate()[0].strip().decode()

files = sorted(glob.glob(os.path.expanduser('~/Downloads/*tri*.dae')))

for filename in files:
    t = timeit.timeit(lambda: collada.Collada(filename), number=3)
    print('%s %20s %.4f %s' % (sys.version[:3], branch, t, os.path.basename(filename)))
```

Results:
```
2.7   tristrips-no-loops 0.0557 collada_tristrips.dae
2.7   tristrips-no-loops 0.0386 duck_trifans.dae
2.7   tristrips-no-loops 0.0227 duck_tristrips.dae
2.7   tristrips-no-loops 0.0088 multi_geometries_multi_tristrips.dae
2.7   tristrips-no-loops 0.0150 multi_geometries_one_trifans.dae
2.7   tristrips-no-loops 0.1163 seymour_plane_trifans.dae

2.7            tristrips 0.0638 collada_tristrips.dae
2.7            tristrips 0.0619 duck_trifans.dae
2.7            tristrips 0.0312 duck_tristrips.dae
2.7            tristrips 0.0111 multi_geometries_multi_tristrips.dae
2.7            tristrips 0.0206 multi_geometries_one_trifans.dae
2.7            tristrips 0.1700 seymour_plane_trifans.dae

3.6   tristrips-no-loops 0.0575 collada_tristrips.dae
3.6   tristrips-no-loops 0.0424 duck_trifans.dae
3.6   tristrips-no-loops 0.0246 duck_tristrips.dae
3.6   tristrips-no-loops 0.0086 multi_geometries_multi_tristrips.dae
3.6   tristrips-no-loops 0.0164 multi_geometries_one_trifans.dae
3.6   tristrips-no-loops 0.1265 seymour_plane_trifans.dae

3.6            tristrips 0.0618 collada_tristrips.dae
3.6            tristrips 0.0614 duck_trifans.dae
3.6            tristrips 0.0294 duck_tristrips.dae
3.6            tristrips 0.0101 multi_geometries_multi_tristrips.dae
3.6            tristrips 0.0206 multi_geometries_one_trifans.dae
3.6            tristrips 0.1752 seymour_plane_trifans.dae
```
Conclusion:
**no-loops** branch up to 30% faster for **trifans** and up to 16% faster for **tristrips** (with 3.6).

Reply:
tests run on Arch Linux, Intel Core i7, 32GB RAM, test files on SSD